Rework `mirrored_supervisor` child ID format

dumbbell opened the pull request:
[TBW]

gomoripeti commented:
I would love to use the testcase `rolling_upgrade` as a template in the future to add rolling upgrade test cases for other components. But how does it work? How does ct know that the "old" nodes have a different version than the "new" one. Aren't they all use the same version (that is the "current" commit whenever it is run)?

michaelklishin commented:
@gomoripeti we have a class of tests known as "mixed version tests". The test itself does not know which node is which, it depends on how you form the cluster. Tests can be skipped when run in mixed version mode.

The rest is up to the test author: run the correct assertions, avoid trying to use features that certain versions do not have, and so on.

dumbbell commented:
> @gomoripeti wrotes:
> I would love to use the testcase `rolling_upgrade` as a template in the future to add rolling upgrade test cases for other components. But how does it work? How does ct know that the "old" nodes have a different version than the "new" one. Aren't they all use the same version (that is the "current" commit whenever it is run)?

common_test doesn't know. It's part of our `rabbitmq_ct_helpers` library. To run tests in mixed-version mode, you can specify the path to a secondary "umbrella" (the term is historic):

```
gmake ct-exchange SECONDARY_UMBRELLA=/path/to/v3.12.x/clone RABBITMQ_FEATURE_FLAGS=
```

`/path/to/v3.12.x/clone` is the root of the Git clone.

If `rabbitmq_ct_helpers` sees this variable set, it will start nodes using both the clone you are running the test from and the secondary umbrella clone in a round-robin fashion:
* node 0 = current clone
* node 1 = secondary umbrella
* node 2 = current clone
* node 3 = secondary umbrella
* …

This new testcase here is built on that assumption.

I recommend that you precompile the secondary umbrella to avoid concurrent builds in `rabbitmq_ct_helpers`. For the context of this testcase:
```
gmake -C /path/to/v3.12.x/clone/deps/rabbitmq_federation test-dist
```

dumbbell commented:
> Tests can be skipped when run in mixed version mode.

They should return `{skip, …}` after they tried to enable the feature flag(s) they depend on. They should **never** **ever** use `rabbit_ct_helpers:is_mixed_version/1` except if there is a very good documented reason.

gomoripeti commented:
thank you both for clarifying how the mixed version tests work, very useful.
sorry for sidetracking comments of this PR.

dumbbell commented:
No problem @gomoripeti! In fact, this prompted me to add a comment in the testcase because indeed, the underlying behavior is not obvious from the testcase itself.

I updated the branch to restore the behavior for the shovel plugin and added a testcase to reproduce what was reported in #9894. I verified manually the new testcase failed when updating from 3.12.7 to 3.12.8.

dumbbell commented:
I think the patch is ready for more review and testing.

@gomoripeti: Could you please give it a try with both the shovel and federation plugins? The new testcases reproduce the scenarios you shared, but still, I would be more confident if you could give it a go.

gomoripeti commented:
I tested upgrading from both 3.12.7 and 3.12.12 to this branch, both on Mnesia and after migration to Khpepri did some modifications to shovels and federation. Couldn't find any issues.

What is the exit plan for removing the code that handles the temporary experimental ids for shovel? I'm not sure if this is a concern, but it sounds like a shovel created on 3.12.12 can survive until the end of time with rolling restarts.

michaelklishin commented:
@gomoripeti right now the plan is to ship 3.13, then we will see what would be the best way forward for 3.12. Likely it will simply support both formats in a simplified manner, that's it.

dumbbell commented:
I’m investigating a transieng failure in CI that I can reproduce locally where a shovel is not restarted as part of the simulated upgrade in the new testsuite.

dumbbell commented:
The problem in question is unrelated to this patch, but something in Khepri. Once CI is happy, I’ll mark this pull request as ready to review.

gomoripeti commented:
> What is the exit plan for removing the code that handles the temporary experimental ids

Yes, I absolutely meant on the long term, and outside of the scope of this PR. I don't think 3.12x and 3.13.x series should do anything more about them.

To be somewhat constructive, a draft idea:
On release series N, when the shovel plugin and `rabbit_shovel_dyn_worker_sup_sup` starts up, it can go through all children and restart (`rabbit_shovel_dyn_worker_sup_sup:stop_child`+`start_child`) those with temporary experimental IDs. Release series N is strictly after 3.13.x. At that point it is guaranteed that even in a mixed version cluster no other node creates temporary IDs (also no need for a feature flag). After an upgrade N - 1 -> N it is guaranteed that a node and sup_sup was restarted at least once, so on release series N + 1 there must be no temporary experimental IDs left and the code that handles them can be removed on N + 1. Going through the children is similar to `cleanup_spec` and most of the time it will find nothing to restart so it should be cheap.

michaelklishin commented:
@gomoripeti honestly, I think a long term solution should start with replacing the `mirrored_supervisor` we have with a Khepri-based version.

Which means we will have a migration to implement and that we will have a lot of design freedom with the new version.

michaelklishin reviewed (approved):
There is some ongoing Khepri-related work that seeks to make the new integration/CT test in this PR non-flaky or much less flaky but otherwise it seems to be good to go.
